traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go

```diff
@@ -0,0 +1,193 @@
+		return nil, err, http.StatusInternalServerError
+	}
+	defer logTx.Commit()
+
+	dsID, certVersion, err := getDSIdAndVersionFromName(db, dsName)
+	if err != nil {
+		return nil, errors.New("querying DS info: " + err.Error()), http.StatusInternalServerError
+	}
+	if dsID == nil || *dsID == 0 {
+		return errors.New("DS id for " + dsName + " was nil or 0"), nil, http.StatusBadRequest
+	}
+	if certVersion == nil || *certVersion == 0 {
+		return errors.New("certificate for " + dsName + " could not be renewed because version was nil or 0"), nil, http.StatusBadRequest
+	}
+
+	if cfg == nil {
+		return nil, errors.New("acme: config was nil"), http.StatusInternalServerError
+	}
+	keyObj, ok, err := riaksvc.GetDeliveryServiceSSLKeysObjV15(dsName, strconv.Itoa(int(*certVersion)), tx, cfg.RiakAuthOptions, cfg.RiakPort)
+	if err != nil {
+		return nil, errors.New("getting ssl keys for xmlId: " + dsName + " and version: " + strconv.Itoa(int(*certVersion)) + " : " + err.Error()), http.StatusInternalServerError
+	}
+	if !ok {
+		return nil, errors.New("no object found for the specified key with xmlId: " + dsName + " and version: " + strconv.Itoa(int(*certVersion))), http.StatusInternalServerError
+	}
+
+	err = base64DecodeCertificate(&keyObj.Certificate)
+	if err != nil {
+		return nil, errors.New("decoding cert for XMLID " + dsName + " : " + err.Error()), http.StatusInternalServerError
+	}
+
+	acmeAccount := GetAcmeAccountConfig(cfg, keyObj.AuthType)
+	if acmeAccount == nil {
+		return nil, errors.New("No acme account information in cdn.conf for " + keyObj.AuthType), http.StatusInternalServerError
+	}
+
+	client, err := GetAcmeClient(acmeAccount, userTx, db)
+	if err != nil {
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: FAILED to add SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+		return nil, errors.New("getting acme client: " + err.Error()), http.StatusInternalServerError
+	}
+
+	renewRequest := certificate.Resource{
+		Certificate: []byte(keyObj.Certificate.Crt),
+	}
+
+	cert, err := client.Certificate.Renew(renewRequest, true, false)
+	if err != nil || cert == nil {
+		log.Errorf("Error obtaining acme certificate: %s", err.Error())
+		return nil, err, http.StatusInternalServerError
+	}
+
+	newCertObj := tc.DeliveryServiceSSLKeys{
+		AuthType:        keyObj.AuthType,
+		CDN:             keyObj.CDN,
+		DeliveryService: keyObj.DeliveryService,
+		Key:             keyObj.DeliveryService,
+		Hostname:        keyObj.Hostname,
+		Version:         keyObj.Version + 1,
+	}
+
+	newCertObj.Certificate = tc.DeliveryServiceSSLKeysCertificate{
+		Crt: string(EncodePEMToLegacyPerlRiakFormat(cert.Certificate)),
+		Key: string(EncodePEMToLegacyPerlRiakFormat(cert.PrivateKey)),
+		CSR: string(EncodePEMToLegacyPerlRiakFormat([]byte("ACME Generated"))),
+	}
+
+	if err := riaksvc.PutDeliveryServiceSSLKeysObj(newCertObj, tx, cfg.RiakAuthOptions, cfg.RiakPort); err != nil {
```

Comment:
post-rebase, once this function takes a `tv trafficvault.TrafficVault`, you would call `tv.PutDeliveryServiceSSLKeys(newCertObj, tx)`